src/Compilers/CSharp/Portable/Binder/Binder_Conversions.cs

```diff
@@ -128,6 +133,98 @@ internal partial class Binder
+            {
+                return MakeBadExpressionForObjectCreation(node, destination, boundInitializerOpt, arguments);
+            }
+
+            BoundExpression result;
+            switch (type.TypeKind)
+            {
+                case TypeKind.Struct:
+                case TypeKind.Class:
+                    result = BindClassCreationExpression(
+                        node,
+                        typeName: type.Name,
+                        typeNode: node,
+                        type: (NamedTypeSymbol)type,
+                        arguments,
+                        diagnostics,
+                        boundInitializerOpt,
+                        forTargetTypedNew: true);
+                    break;
+
+                case TypeKind.TypeParameter:
+                    result = BindTypeParameterCreationExpression(
+                        node,
+                        typeParameter: (TypeParameterSymbol)type,
+                        arguments,
+                        boundInitializerOpt,
+                        diagnostics);
+                    break;
+
+                default:
+                    throw ExceptionUtilities.UnexpectedValue(type.Kind);
+            }
+
+            if (destination.IsNullableType())
+            {
+                result = new BoundConversion(
+                    syntax,
+                    result,
+                    conversion: new Conversion(ConversionKind.ImplicitNullable, Conversion.IdentityUnderlying),
+                    @checked: false,
+                    explicitCastInCode: isCast,
+                    constantValueOpt: null, // A "target-typed new" would never produce a constant.
+                    type: destination);
+            }
+
+            return result;
+        }
+
+        private bool ReportBadTargetType(SyntaxNode syntax, TypeSymbol type, DiagnosticBag diagnostics)
+        {
+            switch (type.TypeKind)
+            {
+                case TypeKind.Array:
+                case TypeKind.Enum:
+                case TypeKind.Delegate:
+                case TypeKind.Interface:
+                    Error(diagnostics, ErrorCode.ERR_BadTargetTypeForNew, syntax, type);
+                    return true;
+
+                case TypeKind.Pointer:
+                    Error(diagnostics, ErrorCode.ERR_UnsafeTypeInObjectCreation, syntax, type);
+                    return true;
+
+                case TypeKind.Dynamic:
+                    Error(diagnostics, ErrorCode.ERR_NoConstructors, syntax, type);
+                    return true;
+
+                case TypeKind.Struct when type.IsTupleType:
+                    Error(diagnostics, ErrorCode.ERR_NewWithTupleTypeSyntax, syntax, type);
+                    return true;
+
+                case TypeKind.Struct:
+                case TypeKind.Class:
+                case TypeKind.TypeParameter:
+                    return false;
+
+                default:
+                    return true;
```

Comment:
>return true; [](start = 20, length = 12)

Is the default case reachable? If not, let's throw an UnexpectedValue exception.

Reply:
Yes, it's reachable, for example `TypeKind.Error`. I'm not sure if other kinds can possibly get here.

Reply:
Thanks for the clarification. I think we should make the `Error` case explicit. Having a complete switch is beneficial, as it catches when we add new cases.

---
In reply to: [206795071](https://github.com/dotnet/roslyn/pull/25196#discussion_r206795071) [](ancestors = 206795071)

Reply:
Done.